libc/common: Add memalign

keith-packard opened the pull request:
Memalign is another name for the posix aligned_alloc function, although it has weaker restrictions on the relationship between the alignment and size.

memalign appears to be used internally by the G++, so we need to provide an implementation of this when using C++, even though it's not part of the Zephyr C library API.

Closes: #57899

keith-packard committed:
libc/common: Add memalign

Memalign is another name for the posix aligned_alloc function, although it
has weaker restrictions on the relationship between the alignment and size.

memalign() is used internally by the libstdc++ when built for 'newlib'
targets (which includes picolibc) instead of aligned_alloc() due to a bug
in gcc, so we need to provide an implementation of this when using that
library, even though it's not part of the Zephyr C library API.

When a fix for the libstdc++ is merged upstream and can be consider a
reasonable dependency for Zephyr, this work-around can be removed.

Closes: #57899

Signed-off-by: Keith Packard <keithp@keithp.com>

cfriedt commented on lib/libc/common/source/stdlib/malloc.c:
```diff
@@ -153,6 +153,14 @@ void *aligned_alloc(size_t alignment, size_t size)
+/* G++ uses this function internally */
+void *memalign(size_t alignment, size_t size)
+{
+	return aligned_alloc(alignment, size);
```
Probably should check if alignment is non-zero and is a power of two.

```
size > 0 && (size & (size - 1)) == 0
```

This might be checked already in zephyr's `k_aligned_alloc()`.

Also, since C is pretty high traffic, it would be great to add a couple of test cases.

One additional nit: please correct the commit message to ensure it says `aligned_alloc()` is part of ISO C. Maybe there was a mixup with the `posix_memalign()` function which is similar.

keith-packard replied:
Good points -- it's not technically valid to implement `memalign` as a direct call to `aligned_alloc` as `aligned_alloc `requires size to be a multiple of alignment while memalign does not. However, the manual for `k_heap_aligned_alloc` reads exactly like the manual for `memalign`, so it is valid to implement both of them with the same underlying code. It would be semantically cleaner to implement `memalign` in terms of `k_heap_aligned_alloc` and then `aligned_alloc` in terms of `memalign`, but as we want to expose only `aligned_alloc` in the Zephyr API, and `memalign` should only be used by the C++ code, the layering chosen here is slightly more efficient. for the desired common case.

If we used a symbol alias instead of a nested function call, then `memalign` would not introduce any additional overhead, but I couldn't find any instance in the Zephyr code using that mechanism.

I can add some comments in the code to make this all a bit more clear.

keith-packard replied:
I've copied the failing code from #57899 into a new test run as part of the existing libcxx tests. That allowed me to do some before/after testing to make sure this patch actually fixes the bug. And I've added more comments in the code to explain why the memalign call is allowed to just directly call aligned_alloc.

galak commented on lib/libc/common/source/stdlib/malloc.c:
```diff
@@ -153,6 +153,14 @@ void *aligned_alloc(size_t alignment, size_t size)
 	return ret;
 }
 
+#ifdef CONFIG_CPP
```
Make this `#if defined(CONFIG_CPP) && defined(__GNUC__)`

keith-packard replied:
that will hit clang as well though, right?

keith-packard replied:
done; we can adjust for clang if necessary.

galak reviewed (dismissed):
tweak ifdef to include `__GNUC__`

keith-packard committed:
tests/libcxx: Test aligned allocation with C++ new operator

Make sure the underlying allocation system can support an allocation
request generated by the new operator which has stricter alignment
requirements than the default. For G++, this ends up using the 'memalign'
function which is not part of any C standard.

Signed-off-by: Keith Packard <keithp@keithp.com>

cfriedt reviewed (dismissed):
Just realized that memalign is nonstandard. I won't block, but likely @stephanosio will have some input.

In any case, I appreciate the effort to improve C++ support :-)

keith-packard commented:
> Just realized that memalign is nonstandard. I won't block, but likely @stephanosio will have some input.

Yup, that's why it's not defined in any header file and is only made available with using G++. It's  provided as part of the G++ ABI requirements for the C library, but it's *not* part of the Zephyr C library API.

> In any case, I appreciate the effort to improve C++ support :-)

I'm constantly surprised what people will do with C++.

stephanosio reviewed (changes requested):
> memalign appears to be used internally by the G++, so we need to provide an implementation of this when using C++, even though it's not part of the Zephyr C library API.

I do not think the right solution is to add the non-standard `memalign` support just for the purpose of appeasing GNU libstdc++, especially when we already support the standard `aligned_alloc`

Moreover, [libstdc++ _should_ be using `aligned_alloc` when it is available](https://github.com/gcc-mirror/gcc/blob/5d85b5d649fff675ff00adcc99371bccf4ef5944/libstdc%2B%2B-v3/libsupc%2B%2B/new_opa.cc#L66-L67):

```c++
namespace __gnu_cxx {
#if _GLIBCXX_HAVE_ALIGNED_ALLOC
using ::aligned_alloc;
[...]
#elif _GLIBCXX_HAVE_MEMALIGN
static inline void*
aligned_alloc (std::size_t al, std::size_t sz)
{
  // Solaris requires al >= sizeof a word and QNX requires >= sizeof(void*)
  // but they both provide posix_memalign, so will use the definition above.
  return memalign (al, sz);
}
[...]
  while ((p = __gnu_cxx::aligned_alloc (align, sz)) == nullptr)
```

Something is causing `_GLIBCXX_HAVE_ALIGNED_ALLOC` to be not defined and the libstdc++ is falling back to using `memalign`, which is a configuration issue and needs to be root-caused and fixed there.

stephanosio commented:
> Something is causing `_GLIBCXX_HAVE_ALIGNED_ALLOC` to be not defined and the libstdc++ is falling back to using `memalign`, which is a configuration issue and needs to be root-caused and fixed there.

I have created https://github.com/zephyrproject-rtos/sdk-ng/issues/675 to track this.

keith-packard commented:
We can fix the SDK, but we cannot fix other gcc-based toolchains. How can we tell when a fixed version of the SDK is being used?

keith-packard commented:
> First of all, we probably should consider upstreaming [zephyrproject-rtos/gcc#22](https://github.com/zephyrproject-rtos/gcc/pull/22), once it is verified working, so that it propagates to all third-party GNU toolchain distributions.

Yup, I'm already planning on doing that. Given the usual gcc merge process, I expect this to take between a week and infinite weeks.

> After that, we can go about this in one of two ways: a) require using the version of Zephyr SDK or a third-party GNU toolchain that includes [the patch](https://github.com/zephyrproject-rtos/gcc/pull/22), or b) provide the solution proposed in this PR as a workaround for the third-party GNU toolchains that do not have the patch.
> 
> IMHO, given that this will not be the last time we encounter issues like this, my preference is towards the option 'a'.

I don't think we can responsibly adopt that solution at this point -- that would rule out using any existing GCC-based c++ toolchain with Zephyr. We can (and should) plan on removing this kludge when the upstream patch is upstream and included in a GCC release.

> > How can we tell when a fixed version of the SDK is being used?
> 
> The only way to do that would be to check the Zephyr SDK version.

Sounds like we need a Kconfig variable driving the inclusion of `memalign()` in the common C library that will be set when using a toolchain which might require it? That seems pretty simple.

stephanosio commented:
> that would rule out using any existing GCC-based c++ toolchain with Zephyr. We can (and should) plan on removing this kludge when the upstream patch is upstream and included in a GCC release.

I am afraid that is an inevitable future if we want to properly support libstdc++ in the future (e.g. synchronisation, threading support), unless the third-party GNU toolchain is specifically built for the `*-zephyr` target.

> Sounds like we need a Kconfig variable driving the inclusion of `memalign()` in the common C library that will be set when using a toolchain which might require it? That seems pretty simple.

Actually, I have another solution in mind. Since we are only adding `memalign()` for the purpose of appeasing GNU libstdc++ (and not for the purpose of supporting it as part of the "Zephyr C API"), we should just add as part of the "GNU libstdc++ integration layer" in `lib/cpp/glibcxx/libcpp-hook.c` or something.

That will allow the `memalign()` to be only defined when `GLIBCXX_LIBCPP` is selected (i.e. when libstdc++ is used), which is what we want.

stephanosio commented:
> Any reason to not stick that test in the common C library malloc.c file instead of the current one?

Because `memalign()` is not supposed to be a libc function that Zephyr supports or is used anywhere in Zephyr. It is analogous to `libc-hooks.c` defining functions like `__retarget_lock_init_recursive` that is used by the Newlib (or Picolibc for that matter).

keith-packard commented:
> Because `memalign()` is not supposed to be a libc function that Zephyr supports or is used anywhere in Zephyr. It is analogous to `libc-hooks.c` defining functions like `__retarget_lock_init_recursive` that is used by the Newlib (or Picolibc for that matter).

I've started implementing this idea and it doesn't seem very pretty -- it depends on configurations that use both the common C library malloc  and libstdc++, so you have a cross-subsystem dependency with either implementation. 

The other issue is that implementing `memalign()` in terms of `aligned_alloc()` is not generally valid as `aligned_alloc()` imposes additional restrictions on the parameters: the size requested must be a multiple of the alignment. Writing `memalign()` in terms of a a pure ISO/IEC C interface will cause significant additional memory usage for applications as allocations will need to be padded to a larger size.

Given these two issues, I'd like you to reconsider this suggestion and place `memalign()` in the common C library malloc implementation making its presence in the resulting binary conditional upon using libstdc++ (via a check for `CONFIG_GLIBCXX_LIBCPP`).

stephanosio commented:
> I've started implementing this idea and it doesn't seem very pretty -- it depends on configurations that use both the common C library malloc and libstdc++, so you have a cross-subsystem dependency with either implementation.

Yes, we will have cross-subsystem dependency either way; under `lib/libc` or `lib/cpp`.

> The other issue is that implementing `memalign()` in terms of `aligned_alloc()` is not generally valid as `aligned_alloc()` imposes additional restrictions on the parameters: the size requested must be a multiple of the alignment. Writing `memalign()` in terms of a a pure ISO/IEC C interface will cause significant additional memory usage for applications as allocations will need to be padded to a larger size.

I thought we were only talking about the common libc `aligned_alloc()`, which is not affected by this, since we need this workaround only when using the common libc. Am I missing something here?

The reason that I say this belongs in `lib/cpp/glibcxx/libcpp-hook.c` (or something equivalent) is that this is strictly for the purpose of supporting the libstdc++ and nothing else (i.e. it is not meant to be a _libc function_, but a libstdc++ support function).

`memalign` in this case is nothing more than an "internal interface" between the libstdc++ and the operating system (as `__retarget_lock_init_recursive` is an internal interface between Newlib and the operating system). The only thing about `memalign` here that makes it as though it should belong in the libc is its name.

That is why I say it would be more logical to put it in the libstdc++ support layer than in a libc.

keith-packard commented:
> I thought we were only talking about the common libc `aligned_alloc()`, which is not affected by this, since we need this workaround only when using the common libc. Am I missing something here?

No, you're correct, but this makes it different from other 'support' code in lib/cpp as it is a linked dependency between the common C library and libstdc++; you need the common C library to provide this interface whenever it is being used by libstdc++. As you cannot use this implementation with any other C library, it feels disingenuous to have it exist outside of the common C library implementation -- it's only within the context of the common C library that this implementation is valid, so I really think it needs to be placed in that code and not someplace where it could conceivably get adopted for use with some other C library where it cannot work.

> The reason that I say this belongs in `lib/cpp/glibcxx/libcpp-hook.c` (or something equivalent) is that this is strictly for the purpose of supporting the libstdc++ and nothing else (i.e. it is not meant to be a _libc function_, but a libstdc++ support function).

But it's a support function that that depends upon internal details about the implementation of `aligned_alloc()` as provided by the common C library, not a support function that exists upon standard C or Zephyr interfaces, which makes it very different from code found in the libc support code (like libc-hooks.c).

> `memalign` in this case is nothing more than an "internal interface" between the libstdc++ and the operating system (as `__retarget_lock_init_recursive` is an internal interface between Newlib and the operating system). The only thing about `memalign` here that makes it as though it should belong in the libc is its name.

Yup, `memalign` will not be visible in header files to Zephyr applications, so they should not have access to it. If it could be implemented purely in terms of standard functions like a ISO/IEC C `aligned_alloc()`, or standard Zephyr functions, then I'd be all for having it placed in lib/cpp/glibcxx.

stephanosio commented:
> But it's a support function that that depends upon internal details about the implementation of `aligned_alloc()` as provided by the common C library, not a support function that exists upon standard C or Zephyr interfaces, which makes it very different from code found in the libc support code (like libc-hooks.c).

As long as we add `#ifdef CONFIG_COMMON_LIBC_MALLOC` (and that should be fine since we only need this when the common libc is used), I think it equally makes sense to place this on the libstdc++ integration layer side as a _libstdc++ support function_ -- for instance, we do not have `aligned_alloc` inside `sys_heap` because it depends on the internal details about how `sys_heap_aligned_alloc` is implemented.

But yeah, it would seem both approaches are almost equally as valid depending on how you look at them, and this is a matter of opinion.

> Yup, `memalign` will not be visible in header files to Zephyr applications, so they should not have access to it. If it could be implemented purely in terms of standard functions like a ISO/IEC C `aligned_alloc()`, or standard Zephyr functions, then I'd be all for having it placed in lib/cpp/glibcxx.

Sure, I am fine with placing `memalign` in the common libc as long as it is not visible in the header files, and the comments make it clear that it is only there for supporting the GNU libstdc++ and is not meant to be used by the application code -- we do not want people submitting PRs saying "`malloc.h` was missing `memalign` function declaration."

keith-packard commented:
> Sure, I am fine with placing `memalign` in the common libc as long as it is not visible in the header files, and the comments make it clear that it is only there for supporting the GNU libstdc++ and is not meant to be used by the application code -- we do not want people submitting PRs saying "`malloc.h` was missing `memalign` function declaration."

Sorry, I hadn't pushed the update I'd made to the comments in the file yesterday which (I hope) address this concern. Thanks for your review; it's always a challenge to know where to stick kludges like this which don't fit the desired abstraction layering.

stephanosio commented on lib/libc/common/source/stdlib/malloc.c:
```diff
@@ -153,6 +153,27 @@ void *aligned_alloc(size_t alignment, size_t size)
+ * Note that aligned_alloc isn't defined to work as a replacement for
+ * memalign as it requires that the size be a multiple of the alignment,
+ * while memalign does not. However, the aligned_alloc implementation here
+ * is just a wrapper around sys_heap_aligned_alloc which doen't have that
```
```suggestion
 * is just a wrapper around sys_heap_aligned_alloc which doesn't have that
```

keith-packard replied:
thanks. fixed.

stephanosio reviewed (commented):
Looks good. Just a minor comment about a typo.

cfriedt reviewed (approved):
No object from me. @galak any further thoughts here?
